tests/unit/test_compress.py

```diff
@@ -0,0 +1,382 @@
+# Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import gzip
+import io
+from copy import deepcopy
+
+import pytest
+
+from botocore.compat import urlencode
+from botocore.compress import COMPRESSION_MAPPING, maybe_compress_request
+from botocore.config import Config
+from tests import mock
+
+
+def _make_op(
+    request_compression=None,
+    has_streaming_input=False,
+    streaming_metadata=None,
+):
+    op = mock.Mock()
+    op.request_compression = request_compression
+    op.has_streaming_input = has_streaming_input
+    if streaming_metadata is not None:
+        streaming_shape = mock.Mock()
+        streaming_shape.metadata = streaming_metadata
+        op.get_streaming_input.return_value = streaming_shape
+    return op
+
+
+OP_NO_COMPRESSION = _make_op()
+OP_WITH_COMPRESSION = _make_op({'encodings': ['gzip']})
+OP_UNKNOWN_COMPRESSION = _make_op({'encodings': ['foo']})
+OP_MULTIPLE_COMPRESSIONS = _make_op({'encodings': ['gzip', 'foo']})
+STREAMING_OP_WITH_COMPRESSION = _make_op(
+    {'encodings': ['gzip']},
+    True,
+    {},
+)
+STREAMING_OP_WITH_COMPRESSION_REQUIRES_LENGTH = _make_op(
+    {'encodings': ['gzip']},
+    True,
+    {'requiresLength': True},
+)
+REQUEST_BODY = (
+    b'Action=PutMetricData&Version=2010-08-01&Namespace=Namespace'
+    b'&MetricData.member.1.MetricName=metric&MetricData.member.1.Unit=Bytes'
+    b'&MetricData.member.1.Value=128'
+)
+
+REQUEST_BODY_STRING = REQUEST_BODY.decode('utf-8')
+
+DEFAULT_COMPRESSION_CONFIG = Config(
+    disable_request_compression=False,
+    request_min_compression_size_bytes=10420,
+)
```

Comment:
Its labeled 'default' because these are what the values will default to from the config provider if not provided by a customer. However, since we are not initializing a client here, we need to write them ourselves. To prevent the actual defaults and these tests from ever getting out of sync, I could do something like this.

```
from botocore.configprovider import BOTOCORE_DEFAUT_SESSION_VARIABLES

default_disable_request_compression = BOTOCORE_DEFAUT_SESSION_VARIABLES['disable_request_compression'[2]
default_request_min_compression_size_bytes = BOTOCORE_DEFAULT_SESSION_VARIABLES['request_min_compression_size_bytes'][2]

DEFAULT_COMPRESSION_CONFIG = Config(
    disable_request_compression=default_disable_request_compression,
    request_min_compression_size_bytes=default_request_min_compression_size_bytes,
)

Reply:
I'm also realizing this constant is misspelled. Should we fix this in a separate PR?

Reply:
This should also have been 10240 not 10420 🤦 . It's fixed in the latest commit